frame/chainrelay/eth/offchain/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+			}
+		}
+	}
+	}
+}
+
+fn remove_trascation_and_uncle(r: &mut Vec<u8>) {
+	let mut pr = 1330;
+	for i in 1330..1394 {
+		if r[i] == 91u8 {
+			pr = i;
+			break;
+		}
+	}
+	let mut tail = r.split_off(pr - 16);
+	if tail[tail.len() - 103 - 68 * 2 - 1 - 1] == 93u8 {
+		tail = tail.split_off(tail.len() - 103 - 68 * 2 - 1);
+		tail.split_off(tail.len() - 15 - 68 * 2 - 1)
+	} else if tail[tail.len() - 103 - 68 - 1] == 93u8 {
+		tail = tail.split_off(tail.len() - 103 - 68);
+		tail.split_off(tail.len() - 15 - 68)
+	} else {
+		tail = tail.split_off(tail.len() - 103);
+		tail.split_off(tail.len() - 15)
+	};
+	r.append(&mut tail);
+	r.push(125u8);
+	r.push(125u8);
+}
+
+fn json_request(raw_url: &Vec<u8>, api_type: EthScanAPI) -> Result<JsonValue> {
+	let url = core::str::from_utf8(raw_url).map_err(|_| "url decode error")?;
+	debug::trace!("[eth-offchain] request: {:?}", url);
+
+	let pending = http::Request::get(&url)
+		.send()
+		.map_err(|_| "Error in sending http GET request")?;
+
+	let mut response = pending.wait().map_err(|_| "Error in waiting http response back")?;
+
+	let mut retry_time = 0;
+	let mut r: Vec<u8>;
+	loop {
+		if response.code != 200 {
+			if retry_time == 3 {
+				debug::warn!("Unexpected status code: {}", response.code);
+				return Err("Non-200 status code returned from http request");
+			}
+			response = http::Request::get(&url)
+				.send()
+				.map_err(|_| "Error in sending http GET request")?
+				.wait()
+				.map_err(|_| "Error in waiting http response back")?;
+			retry_time += 1;
+			debug::info!("[eth-offchain] retry {} times", retry_time);
+		} else {
+			r = response.body().collect::<Vec<u8>>();
+			if r[0] == 123u8 {
+				break;
+			}
+		}
+		// TODO: figure out how to sleep in no-std here
+	}
+
+	debug::trace!("[eth-offchain] response: {:?}", core::str::from_utf8(&r));
+
+	let json_val: JsonValue = match api_type {
+		EthScanAPI::GetBlockByNumber => {
+			if r.len() < 1362 {
+				debug::warn!("[eth-offchain] response: {:?}", core::str::from_utf8(&r));
+				return Err("unexpected api response");
+			}
+			remove_trascation_and_uncle(&mut r);
+			// get the result part
+			simple_json::parse_json(
+				&core::str::from_utf8(&r[33..r.len() - 1]).map_err(|_| "result part cannot convert to string")?,
+			)
+			.map_err(|_| "JSON parsing error")?
+		}
+		_ => simple_json::parse_json(&core::str::from_utf8(&r).map_err(|_| "JSON result cannot convert to string")?)
+			.map_err(|_| "JSON parsing error")?,
+	};
+
+	Ok(json_val)
+}
+
+fn hexstr_padding(width: usize, content: AllocString) -> AllocString {
+	if content.len() < width {
+		let mut output: Vec<u8> = Vec::new();
+		for _ in 0..(width - content.len() + 2) {
+			output.push(48);
+		}
+		output.append(&mut content.into_bytes()[2..].to_vec());
+		return output.iter().map(|u| *u as char).collect::<AllocString>();
+	}
+	content
+}
+
+fn build_eth_header(number: u64, block_info: JsonValue) -> Result<EthHeader> {
```

Comment:
How about build a off-chain api wrap service, and do the parse stuffs off chain?

Reply:
This worker gets the block info now relies on Ethscan's API, so it already wraps a web service.  

I think the way to avoid parsing a string into bytes is to use more low-level client,  and there is still no grpc/socket client API in the off-chain worker.  So I think parsing is not avoidable.

By the way, the response format of Ethscan is a JSON with sorted fields, and also is the same as the Ethereum gateway of Cloudflare.  And the worker is easily integrated with multiple sources and will be implemented later to enhance the robust.